File: app/game/scoring.py
from dataclasses import dataclass

from app.game.board import PlayerBoard
from app.game.hand import HandRank, HandValue
from app.game.rules import PINEAPPLE_OFC, Ruleset


@dataclass(frozen=True, slots=True)
class HeadToHeadDetail:
    """a 기준 매치업 분해. 줄별로 라인 ±1과 royalty가 분리 노출됨."""

    top_line: int
    middle_line: int
    bottom_line: int
    top_royalty_a: int
    top_royalty_b: int
    middle_royalty_a: int
    middle_royalty_b: int
    bottom_royalty_a: int
    bottom_royalty_b: int
    scoop: int
    foul_a: bool
    foul_b: bool
    total: int


def royalty_top(board: PlayerBoard, rules: Ruleset = PINEAPPLE_OFC) -> int:
    if board.is_foul or len(board.top) != rules.top_size:
        return 0
    hv = board.top_value()
    if hv.rank == HandRank.THREE_OF_A_KIND:
        return rules.top_trips_royalty.get(hv.high_card, 0)
    if hv.rank == HandRank.ONE_PAIR:
        return rules.top_pair_royalty.get(hv.high_card, 0)
    return 0


def royalty_middle(board: PlayerBoard, rules: Ruleset = PINEAPPLE_OFC) -> int:
    if board.is_foul or len(board.middle) != rules.middle_size:
        return 0
    return rules.middle_royalty.get(board.middle_value().rank, 0)


def royalty_bottom(board: PlayerBoard, rules: Ruleset = PINEAPPLE_OFC) -> int:
    if board.is_foul or len(board.bottom) != rules.bottom_size:
        return 0
    return rules.bottom_royalty.get(board.bottom_value().rank, 0)


def total_royalty(board: PlayerBoard, rules: Ruleset = PINEAPPLE_OFC) -> int:
    return (
        royalty_top(board, rules)
        + royalty_middle(board, rules)
        + royalty_bottom(board, rules)
    )


def head_to_head_detail(
    a: PlayerBoard, b: PlayerBoard, rules: Ruleset = PINEAPPLE_OFC
) -> HeadToHeadDetail:
    """a 기준 매치업 분해.

    Foul은 자동 3패. 줄별 royalty는 foul인 쪽이 0. Scoop 보너스는
    line_sum이 ±3일 때 적용 — foul 매치업도 자동 sweep으로 ±3.
    """
    foul_a, foul_b = a.is_foul, b.is_foul

    top_roy_a = 0 if foul_a else royalty_top(a, rules)
    middle_roy_a = 0 if foul_a else royalty_middle(a, rules)
    bottom_roy_a = 0 if foul_a else royalty_bottom(a, rules)
    top_roy_b = 0 if foul_b else royalty_top(b, rules)
    middle_roy_b = 0 if foul_b else royalty_middle(b, rules)
    bottom_roy_b = 0 if foul_b else royalty_bottom(b, rules)

    if foul_a and foul_b:
        top_line = middle_line = bottom_line = 0
    elif foul_a:
        top_line = middle_line = bottom_line = -1
    elif foul_b:
        top_line = middle_line = bottom_line = 1
    else:
        top_line = _cmp(a.top_value(), b.top_value())
        middle_line = _cmp(a.middle_value(), b.middle_value())
        bottom_line = _cmp(a.bottom_value(), b.bottom_value())

    line_sum = top_line + middle_line + bottom_line
    scoop = (
        rules.scoop_bonus
        if line_sum == 3
        else -rules.scoop_bonus
        if line_sum == -3
        else 0
    )
    royalty_diff = (top_roy_a + middle_roy_a + bottom_roy_a) - (
        top_roy_b + middle_roy_b + bottom_roy_b
    )
    total = line_sum + scoop + royalty_diff

    return HeadToHeadDetail(
        top_line=top_line,
        middle_line=middle_line,
        bottom_line=bottom_line,
        top_royalty_a=top_roy_a,
        top_royalty_b=top_roy_b,
        middle_royalty_a=middle_roy_a,
        middle_royalty_b=middle_roy_b,
        bottom_royalty_a=bottom_roy_a,
        bottom_royalty_b=bottom_roy_b,
        scoop=scoop,
        foul_a=foul_a,
        foul_b=foul_b,
        total=total,
    )


def head_to_head(a: PlayerBoard, b: PlayerBoard, rules: Ruleset = PINEAPPLE_OFC) -> int:
    """a 기준 순점수 반환 (b에게서 a가 얻는 점수)."""
    return head_to_head_detail(a, b, rules).total


def calculate_scores(
    boards: list[PlayerBoard], rules: Ruleset = PINEAPPLE_OFC
) -> list[int]:
    """n명 플레이어의 최종 점수를 계산 (제로섬). boards[i]가 i번 플레이어."""
    n = len(boards)
    scores = [0] * n
    for i in range(n):
        for j in range(i + 1, n):
            delta = head_to_head(boards[i], boards[j], rules)
            scores[i] += delta
            scores[j] -= delta
    return scores


def _cmp(a: HandValue, b: HandValue) -> int:
    if a > b:
        return 1
    if a < b:
        return -1
    return 0

File: frontend/src/lib/scoring.ts
// 백엔드 app/game/scoring.py의 head_to_head_detail 미러.
// 변경 시 양쪽 함께 수정.

import type { Card, Matchup } from "../api/types";
import { evaluate, type HandValue, isFoulBoard } from "./handEval";
import { royaltyBottom, royaltyMiddle, royaltyTop } from "./royalty";

const SCOOP_BONUS = 3;

interface Board {
  top: Card[];
  middle: Card[];
  bottom: Card[];
}

function compareHand(a: HandValue, b: HandValue): number {
  if (a.rank !== b.rank) return a.rank - b.rank;
  const n = Math.max(a.tiebreakers.length, b.tiebreakers.length);
  for (let i = 0; i < n; i++) {
    const x = a.tiebreakers[i] ?? 0;
    const y = b.tiebreakers[i] ?? 0;
    if (x !== y) return x - y;
  }
  return 0;
}

function cmpLine(a: Card[], b: Card[]): number {
  const d = compareHand(evaluate(a), evaluate(b));
  if (d > 0) return 1;
  if (d < 0) return -1;
  return 0;
}

export function headToHeadMatchup(
  aId: string,
  bId: string,
  a: Board,
  b: Board,
): Matchup {
  const foulA = isFoulBoard(a.top, a.middle, a.bottom);
  const foulB = isFoulBoard(b.top, b.middle, b.bottom);

  const topRoyA = foulA ? 0 : royaltyTop(a.top, false);
  const midRoyA = foulA ? 0 : royaltyMiddle(a.middle, false);
  const botRoyA = foulA ? 0 : royaltyBottom(a.bottom, false);
  const topRoyB = foulB ? 0 : royaltyTop(b.top, false);
  const midRoyB = foulB ? 0 : royaltyMiddle(b.middle, false);
  const botRoyB = foulB ? 0 : royaltyBottom(b.bottom, false);

  let topLine: number, middleLine: number, bottomLine: number;
  if (foulA && foulB) {
    topLine = middleLine = bottomLine = 0;
  } else if (foulA) {
    topLine = middleLine = bottomLine = -1;
  } else if (foulB) {
    topLine = middleLine = bottomLine = 1;
  } else {
    topLine = cmpLine(a.top, b.top);
    middleLine = cmpLine(a.middle, b.middle);
    bottomLine = cmpLine(a.bottom, b.bottom);
  }

  const lineSum = topLine + middleLine + bottomLine;
  const scoop = lineSum === 3 ? SCOOP_BONUS : lineSum === -3 ? -SCOOP_BONUS : 0;
  const royaltyDiff =
    topRoyA + midRoyA + botRoyA - (topRoyB + midRoyB + botRoyB);
  const total = lineSum + scoop + royaltyDiff;

  return {
    a_id: aId,
    b_id: bId,
    top_line_a: topLine,
    top_royalty_a: topRoyA,
    top_royalty_b: topRoyB,
    middle_line_a: middleLine,
    middle_royalty_a: midRoyA,
    middle_royalty_b: midRoyB,
    bottom_line_a: bottomLine,
    bottom_royalty_a: botRoyA,
    bottom_royalty_b: botRoyB,
    scoop_a: scoop,
    foul_a: foulA,
    foul_b: foulB,
    total_a: total,
  };
}
